### PR TITLE
[SHOPWARE-7] different twitter timeline types

### PR DESCRIPTION
**Pull request by BrocksiNet**
SHOPWARE-7

### CODE REVIEW

File: src/Resources/app/administration/src/module/sw-cms/elements/twitter-timeline/component/index.js [hunk 1]
```diff
@@ -27,22 +45,49 @@ Component.register('sw-cms-el-twitter-timeline', {
...
-                return 'https://twitter.com/' + this.element.config.twitterHandle.value;
+                return 'https://twitter.com/' + elemData.twitterHandle.value;
             }
         },
 
-        twitterText() {
+        setTwitterHref(twitterHref) {
+            const elemData = this.element.config;
+            if (elemData && elemData.twitterHref) {
+                this.element.config.twitterHref.value = twitterHref;
```
Comment (nekwo12) [inline src/Resources/app/administration/src/module/sw-cms/elements/twitter-timeline/component/index.js]:
```suggestion
                elemData.twitterHref.value = twitterHref;
```

File: src/Resources/app/administration/src/module/sw-cms/elements/twitter-timeline/component/index.js [hunk 2]
```diff
@@ -27,22 +45,49 @@ Component.register('sw-cms-el-twitter-timeline', {
...
+                if (elemData.timelineType.value === 'lists' && elemData.timelineList) {
+                    return twitterHandle + '/lists/' + elemData.timelineList.value;
+                }
+            }
+        },
 
+        twitterText() {
+            const elemData = this.element.config;
             if (elemData && elemData.twitterHandle) {
                 return 'Tweets by ' + this.element.config.twitterHandle.value;
```
Comment (nekwo12) [inline src/Resources/app/administration/src/module/sw-cms/elements/twitter-timeline/component/index.js]:
```suggestion
                return 'Tweets by ' + elemData.twitterHandle.value;
```

File: src/Resources/app/administration/src/module/sw-cms/elements/twitter-timeline/config/sw-cms-el-config-twitter-timeline.html.twig [hunk 3]
```diff
@@ -1,11 +1,43 @@
...
+                  :placeholder="$tc('sw-cms.elements.twitterHandle.placeHolder')">
         </sw-field>
-    </div>
+    {% endblock %}
+
+    {% block sw_cms_element_twitter_timeline_config_timeline_type %}
+    <sw-select-field v-model="element.config.timelineType.value"
+                     :label="$tc('sw-cms.elements.twitterTimeline.type.label')"
+                     :placeholder="$tc('sw-cms.elements.twitterTimeline.type.placeHolder')">
+        <option value="profile">{{ $tc('sw-cms.elements.twitterTimeline.type.profile') }}</option>
```
Comment (nekwo12) [inline src/Resources/app/administration/src/module/sw-cms/elements/twitter-timeline/config/sw-cms-el-config-twitter-timeline.html.twig]: Options can be put inside `src/Resources/app/administration/src/module/sw-cms/elements/twitter-timeline/config/index.js`

like this:
```javascript
data() {
        return {
            listTypes: [
                {
                    id: 'option1',
                    name: "label1",
                },
                {
                    id: 'option2',
                    name: "label2",
                },
            ],
        };
    },
```

File: src/Resources/app/administration/src/module/sw-cms/elements/twitter-timeline/config/sw-cms-el-config-twitter-timeline.html.twig [hunk 4]
```diff
@@ -1,11 +1,43 @@
...
+        <option value="profileLikes">{{ $tc('sw-cms.elements.twitterTimeline.type.profileLikes') }}</option>
+        <option value="collections">{{ $tc('sw-cms.elements.twitterTimeline.type.collections') }}</option>
+        <option value="lists">{{ $tc('sw-cms.elements.twitterTimeline.type.lists') }}</option>
+    </sw-select-field>
+    {% endblock %}
+
+    {% block sw_cms_element_twitter_timeline_config_timeline_list %}
+    <sw-field class="timeline-list-field"
+              v-if="element.config.timelineType && element.config.timelineType.value === 'lists'"
```
Comment (nekwo12) [inline src/Resources/app/administration/src/module/sw-cms/elements/twitter-timeline/config/sw-cms-el-config-twitter-timeline.html.twig]: The condition logic should be a function and inside `src/Resources/app/administration/src/module/sw-cms/elements/twitter-timeline/config/index.js`

File: src/Resources/app/administration/src/module/sw-cms/elements/twitter-timeline/config/sw-cms-el-config-twitter-timeline.html.twig [hunk 5]
```diff
@@ -1,11 +1,43 @@
...
+              v-model="element.config.timelineList.value"
+              type="text"
+              :label="$tc('sw-cms.elements.timelineList.label')"
+              :placeholder="$tc('sw-cms.elements.timelineList.placeHolder')">
+    </sw-field>
+    {% endblock %}
+
+    {% block sw_cms_element_twitter_timeline_config_timeline_collection %}
+    <sw-field class="timeline-collection-field"
+              v-if="element.config.timelineType && element.config.timelineType.value === 'collections'"
```
Comment (nekwo12) [inline src/Resources/app/administration/src/module/sw-cms/elements/twitter-timeline/config/sw-cms-el-config-twitter-timeline.html.twig]: See comment about condition

**Review by nekwo12 — CHANGES_REQUESTED** [review]
Changes look good so far just need to finalize german translations and take a look at a couple of suggestions

File: src/Resources/app/administration/src/module/sw-cms/elements/twitter-timeline/config/sw-cms-el-config-twitter-timeline.html.twig (hunk 4 above)
Comment (BrocksiNet) [reply src/Resources/app/administration/src/module/sw-cms/elements/twitter-timeline/config/sw-cms-el-config-twitter-timeline.html.twig]: I had it before like that, but it is so an easy condition I decided to have it direct in the template

File: src/Resources/app/administration/src/module/sw-cms/elements/twitter-timeline/config/sw-cms-el-config-twitter-timeline.html.twig (hunk 5 above)
Comment (BrocksiNet) [reply src/Resources/app/administration/src/module/sw-cms/elements/twitter-timeline/config/sw-cms-el-config-twitter-timeline.html.twig]: See comment above ;-)

File: src/Resources/app/administration/src/module/sw-cms/elements/twitter-timeline/config/sw-cms-el-config-twitter-timeline.html.twig (hunk 3 above)
Comment (BrocksiNet) [reply src/Resources/app/administration/src/module/sw-cms/elements/twitter-timeline/config/sw-cms-el-config-twitter-timeline.html.twig]: This is intressting. Funny in SW6 core they also add options in template.

Comment (BrocksiNet) [reply src/Resources/app/administration/src/module/sw-cms/elements/twitter-timeline/config/sw-cms-el-config-twitter-timeline.html.twig]: Hmm I was think also about this, but then I have to add more code just for some option that will not change that often. So I would prefer to leave it small and everything in the template.

You need to do something like this:
https://vuejs.org/v2/guide/forms.html#Select

**Comment by BrocksiNet** [comment]
I solved all comments. PR is updated.

File: README.md [hunk 6]
```diff
@@ -18,8 +18,8 @@ All possible twitter elements you can see at [publish.twitter.com](https://publi
 * Shopware 6.1. at least
 * PHP 7.2 at least
 ## Screens
-Twitter timeline frontend  
+###Twitter timeline frontend
```
Comment (nekwo12) [inline README.md]: Formatting should be consistent with line preceding line 20 `## Screens` (notice the whitespace after heading notation)

Generally speaking the file should be formatted differently. Make use of spacing to make it more readable. If you look at this file in the raw format everything looks cramped.

File: README.md [hunk 7]
```diff
@@ -18,8 +18,8 @@ All possible twitter elements you can see at [publish.twitter.com](https://publi
 * Shopware 6.1. at least
 * PHP 7.2 at least
 ## Screens
-Twitter timeline frontend  
+###Twitter timeline frontend  
 ![Twitter timeline in frontend](https://raw.githubusercontent.com/flagbit/shopware6-twitter/master/docs/images/shopware-6-twitter-timeline-frontend.png "Twitter timeline in frontend")
-Twitter timeline backend  
-![Twitter timeline backend configuration](https://raw.githubusercontent.com/flagbit/shopware6-twitter/master/docs/images/shopware-6-twitter-timeline-backend.png "Twitter timeline backend configuration")
+###Twitter timeline backend  
+![Twitter timeline backend configuration](https://raw.githubusercontent.com/flagbit/shopware6-twitter/97b07d18b75e5d0cdc1623b2c3c3b80f396c1b29/docs/images/shopware-6-twitter-timeline-backend.png "Twitter timeline backend configuration")
```
Comment (nekwo12) [inline README.md]:
```suggestion
![Twitter timeline backend configuration](https://raw.githubusercontent.com/flagbit/shopware6-twitter/master/docs/images/shopware-6-twitter-timeline-backend.png "Twitter timeline backend configuration")
```
Should link to master as well

File: src/Resources/app/administration/src/module/sw-cms/elements/twitter-timeline/component/index.js [hunk 8]
```diff
@@ -27,22 +45,49 @@ Component.register('sw-cms-el-twitter-timeline', {
     methods: {
         twitterHandle() {
             const elemData = this.element.config;
-
             if (elemData && elemData.twitterHandle) {
-                return 'https://twitter.com/' + this.element.config.twitterHandle.value;
+                return 'https://twitter.com/' + elemData.twitterHandle.value;
```
Comment (nekwo12) [inline src/Resources/app/administration/src/module/sw-cms/elements/twitter-timeline/component/index.js]: Missing return statement outside of condition

File: src/Resources/app/administration/src/module/sw-cms/elements/twitter-timeline/component/index.js [hunk 9]
```diff
@@ -27,22 +45,49 @@ Component.register('sw-cms-el-twitter-timeline', {
     methods: {
         twitterHandle() {
```
Comment (nekwo12) [inline src/Resources/app/administration/src/module/sw-cms/elements/twitter-timeline/component/index.js]: Method should have a verb to indicate that is does something

File: src/Resources/app/administration/src/module/sw-cms/elements/twitter-timeline/component/index.js [hunk 10]
```diff
@@ -27,22 +45,49 @@ Component.register('sw-cms-el-twitter-timeline', {
...
 
-        twitterText() {
+        setTwitterHref(twitterHref) {
+            const elemData = this.element.config;
+            if (elemData && elemData.twitterHref) {
+                elemData.twitterHref.value = twitterHref;
+            }
+        },
+
+        twitterHref() {
```
Comment (nekwo12) [inline src/Resources/app/administration/src/module/sw-cms/elements/twitter-timeline/component/index.js]: Method should have a verb to indicate that it does something

File: src/Resources/app/administration/src/module/sw-cms/elements/twitter-timeline/component/index.js [hunk 11]
```diff
@@ -27,22 +45,49 @@ Component.register('sw-cms-el-twitter-timeline', {
...
+                    return twitterHandle + '/timelines/' + elemData.timelineCollection.value;
+                }
+
+                if (elemData.timelineType.value === 'lists' && elemData.timelineList) {
+                    return twitterHandle + '/lists/' + elemData.timelineList.value;
+                }
+            }
+        },
 
+        twitterText() {
```
Comment (nekwo12) [inline src/Resources/app/administration/src/module/sw-cms/elements/twitter-timeline/component/index.js]: Method should have a verb to indicate that it does something

File: src/Resources/app/administration/src/module/sw-cms/elements/twitter-timeline/component/index.js [hunk 12]
```diff
@@ -27,22 +45,49 @@ Component.register('sw-cms-el-twitter-timeline', {
...
+                    return twitterHandle + '/likes';
+                }
+
+                if (elemData.timelineType.value === 'collections' && elemData.timelineCollection) {
+                    return twitterHandle + '/timelines/' + elemData.timelineCollection.value;
+                }
+
+                if (elemData.timelineType.value === 'lists' && elemData.timelineList) {
+                    return twitterHandle + '/lists/' + elemData.timelineList.value;
+                }
```
Comment (nekwo12) [inline src/Resources/app/administration/src/module/sw-cms/elements/twitter-timeline/component/index.js:77]: Missing return statement outside of if conditions

File: src/Resources/app/administration/src/module/sw-cms/elements/twitter-timeline/component/index.js [hunk 13]
```diff
@@ -27,22 +45,49 @@ Component.register('sw-cms-el-twitter-timeline', {
...
+            }
+        },
 
+        twitterText() {
+            const elemData = this.element.config;
             if (elemData && elemData.twitterHandle) {
-                return 'Tweets by ' + this.element.config.twitterHandle.value;
+                return 'Tweets by ' + elemData.twitterHandle.value;
```
Comment (nekwo12) [inline src/Resources/app/administration/src/module/sw-cms/elements/twitter-timeline/component/index.js:86]: Missing return statement outside of if conditions

File: src/Resources/app/administration/src/module/sw-cms/snippet/de-DE.json [hunk 14]
```diff
@@ -1,8 +1,28 @@
 {
   "sw-cms": {
     "elements": {
-      "customTwitterTimelineElement": {
-        "label": "Twitter Timeline"
+      "twitterTimeline": {
+        "label": "Twitter Timeline",
```
Comment (nekwo12) [inline src/Resources/app/administration/src/module/sw-cms/snippet/de-DE.json]: placeholder missing (see en-GB)

**Review by nekwo12 — CHANGES_REQUESTED** [review]
Overall code looks good to me :+1: 
Only a couple of adjustments have to be made. Missing Translations for example.

File: README.md [hunk 15]
```diff
@@ -14,12 +15,15 @@ All possible twitter elements you can see at [publish.twitter.com](https://publi
...
 > Legend: Implemented :white_check_mark:, Under construction :construction:
+
 ## Requirements
 * Shopware 6.1. at least
 * PHP 7.2 at least
+
 ## Screens
-###Twitter timeline frontend  
+
+###Twitter timeline frontend
```
Comment (nekwo12) [inline README.md]:
```suggestion
### Twitter timeline frontend
```

File: README.md [hunk 16]
```diff
@@ -14,12 +15,15 @@ All possible twitter elements you can see at [publish.twitter.com](https://publi
...
 ![Twitter timeline in frontend](https://raw.githubusercontent.com/flagbit/shopware6-twitter/master/docs/images/shopware-6-twitter-timeline-frontend.png "Twitter timeline in frontend")
-###Twitter timeline backend  
-![Twitter timeline backend configuration](https://raw.githubusercontent.com/flagbit/shopware6-twitter/97b07d18b75e5d0cdc1623b2c3c3b80f396c1b29/docs/images/shopware-6-twitter-timeline-backend.png "Twitter timeline backend configuration")
 
+###Twitter timeline backend
```
Comment (nekwo12) [inline README.md]:
```suggestion
### Twitter timeline backend
```

File: src/Resources/app/administration/src/module/sw-cms/elements/twitter-timeline/component/index.js [hunk 17]
```diff
@@ -77,13 +76,17 @@ Component.register('sw-cms-el-twitter-timeline', {
                     return twitterHandle + '/lists/' + elemData.timelineList.value;
                 }
             }
+
+            return twitterHandle; // in default profile will be returned
```
Comment (nekwo12) [inline src/Resources/app/administration/src/module/sw-cms/elements/twitter-timeline/component/index.js]:
```suggestion
            return twitterHandle; // default profile will be returned
```

**Pull request merged**